### PR TITLE
RSDK-8377: Implement With() equivalent function for RDK loggers

### PR DESCRIPTION
**Pull request by danielbotros**
### Description
RSDK-8377: Implement With() equivalent function for RDK loggers
* Any loggers that call `With()` will get lost by the logger registry as they get converted to `zap.SugaredLogger`s. 
* This PR adds similar structured logging functionality as `With()` to the RDK logger with the function `WithFields()` so that we can keep track of these loggers, and it is just a beneficial feature to have.  
  * Removes usages of `With()` and removed from the logger interface
* Converting these old uses of `With()`, especially in `goutils` to use existing structured logging functions quickly became difficult as they were multiple calls on the same logger and it was very hard to figure out which loggers / log lines should have these key-value pairs + making these key's values accessible to the various log lines. 
  * `goutils` side of changes found [here](https://github.com/viamrobotics/goutils/pull/316).

### Testing
* A test `TestLoggingWithFields` that tests adding key value pairs with various types of keys and values on the various logging methods.

### CODE REVIEW

File: logging/logger.go [hunk 1]
```diff
@@ -46,7 +47,6 @@ type ZapCompatibleLogger interface {
 	Level() zapcore.Level
 	Named(name string) *zap.SugaredLogger
 	Sync() error
-	With(args ...interface{}) *zap.SugaredLogger
```
Comment (dgottlieb) [inline logging/logger.go:49]: I assume this means nothing in goutils was using this version of the With API?

And I see us only using the ZapCompatibleLogger for robot client connections. So removing the restrictions here seems totally safe. Just speaking out loud.

File: logging/impl.go [hunk 2]
```diff
@@ -98,14 +100,34 @@ func (imp *impl) Sync() error {
...
+		case string:
+			keyStr = key
+		case fmt.Stringer:
+			keyStr = key.String()
+		default:
+			continue
+		}
+
+		if keyIdx+1 < len(args) {
+			imp.logFields = append(imp.logFields, zap.Any(keyStr, args[keyIdx+1]))
```
Comment (dgottlieb) [inline logging/impl.go:122]: A detail that I think has no impact -- if a user calls:
```
logger.WithFields("a", 1)
logger.WithFields("a", 2)
```
I expect this code will add 2 items to the log entry. And I suppose a zap logger would only add `"a": 2`?

I'm sure eyeballing to verify that goutils does not do this is trivial. But might be worth a callout at the top of this function. So a reader knows this was an intentional choice and that "last one wins" isn't considered incorrect.

**Review by dgottlieb — APPROVED** [review]

File: logging/impl.go [hunk 3]
```diff
@@ -98,14 +100,34 @@ func (imp *impl) Sync() error {
...
-func (imp *impl) With(args ...interface{}) *zap.SugaredLogger {
-	return imp.AsZap().With(args...)
-}
-
 func (imp *impl) WithOptions(opts ...zap.Option) *zap.SugaredLogger {
 	return imp.AsZap().WithOptions(opts...)
 }
 
+func (imp *impl) WithFields(args ...interface{}) {
+	for keyIdx := 0; keyIdx < len(args); keyIdx += 2 {
```
Comment (benjirewis) [inline logging/impl.go:108]: Was this copied from `zap` code? Mind leaving a permalink here if so?

File: logging/impl_test.go [hunk 4]
```diff
@@ -298,3 +300,146 @@ func TestSublogger(t *testing.T) {
 	assertLogMatches(t, notStdout,
 		`2023-10-30T09:12:09.459Z	INFO	impl.sub	logging/impl_test.go:67	info log`)
 }
+
+func TestLoggingWithFields(t *testing.T) {
```
Comment (benjirewis) [inline logging/impl_test.go:304]: Excellent job on these tests 🧑‍🔧 . Some interesting edge cases covered.

File: logging/logger.go (hunk 1 above)
Comment (benjirewis) [reply logging/logger.go:49]: > this means nothing in goutils was using this version of the With API

I believe there were a few usages of `With` from goutils, but judging from your comment on the goutils PR, maybe this note is resolved?

Comment (dgottlieb) [reply logging/logger.go:49]: Yep!

File: logging/impl.go (hunk 2 above)
Comment (danielbotros) [reply logging/impl.go:126]: this is a good point still since it does slightly deviate from `zap.SugaredLogger` behavior so I will make sure to document

File: logging/impl.go (hunk 3 above)
Comment (danielbotros) [reply logging/impl.go:112]: this was copied and slightly modified from `formatw()` in this file

File: logging/impl_test.go (hunk 4 above)
Comment (danielbotros) [reply logging/impl_test.go:304]: I stole them above from a different test lol

**Review by ethanlook — COMMENTED** [review]
Looks like you accidentally included a bunch of extra commits from main which meant that this PR auto-requested a few of us as code owners of `web/frontend`. Do you mind reverting those extra commits? Thanks!

File: logging/impl.go [hunk 5]
```diff
@@ -201,7 +239,7 @@ func (imp *impl) formatw(logLevel Level, traceKey, msg string, keysAndValues ...
 	logEntry.Level = logLevel.AsZap()
 	logEntry.Message = msg
 
-	logEntry.fields = make([]zapcore.Field, 0, len(keysAndValues)/2+1)
+	logEntry.fields = make([]zapcore.Field, 0, len(imp.logFields)+len(keysAndValues)/2+1)
```
Comment (benjirewis) [inline logging/impl.go:242]: [super-nit] Is that math still right? Missing some parens?

File: logging/impl_test.go (hunk 4 above)
Comment (benjirewis) [reply logging/impl_test.go:304]: Quality theft; thanks @dgottlieb (I'm assuming) for the originals.

**Review by benjirewis — APPROVED** [review]
LGTM % my one math comment and the eventual bumping of goutils.

File: logging/impl.go (hunk 5 above)
Comment (danielbotros) [reply logging/impl.go:242]: I think this math is correct because `imp.logFields` is a slice of `zapcore.Field` which themselves are key-value pairs, while `keysAndValues` is a `interface{}` where every other element is a key or value.

File: logging/impl.go [hunk 6]
```diff
@@ -474,3 +509,325 @@ func getCaller() zapcore.EntryCaller {
...
+}
+
+func (imp *implWith) Debug(args ...interface{}) {
+	imp.logger.testHelper()
+	if imp.logger.shouldLog(DEBUG) {
+		entry := imp.logger.format(DEBUG, emptyTraceKey, args...)
+		entry.fields = append(entry.fields, imp.logFields...)
+		imp.Write(entry)
+	}
+}
```
Comment (danielbotros) [inline logging/impl.go:567]: Calling `imp.logger.Debugw(...)` here with `imp.logFields` also works, but this makes it so that the file name / line number don't properly get logged, which is fixed by just copying and slightly modifying implementation of `impl` functions.

**Comment by danielbotros** [comment]
The old implementation had a data race on `imp.logFields` since two goroutines could call `logger.WithFields(...)`, so instead of adding a lock around every `imp.logFields` I changed the implementation to return a new `Logger` type which implements the same interface except always logs itself `logFields` as well. This change requires a small refactor of the goutils `AddFieldsToLogger` which can be found [here](https://github.com/viamrobotics/goutils/pull/318/files).

File: logging/impl.go [hunk 7]
```diff
@@ -32,6 +32,11 @@ type (
 		zapcore.Entry
 		fields []zapcore.Field
 	}
+
+	implWith struct {
+		logger    *impl
```
Comment (benjirewis) [inline logging/impl.go:37]: Can you embed this field so we can leverage method-forwarding to avoid reimplementing all the methods as you have?

Comment (danielbotros) [reply logging/impl.go]: done!

File: go.mod [hunk 8]
```diff
@@ -87,7 +87,7 @@ require (
 	go.uber.org/zap v1.24.0
 	go.viam.com/api v0.1.322
 	go.viam.com/test v1.1.1-0.20220913152726-5da9916c08a2
-	go.viam.com/utils v0.1.87
+	go.viam.com/utils v0.1.88
```
Comment (benjirewis) [inline go.mod:90]: @dgottlieb note that [this update](https://github.com/viamrobotics/goutils/releases/tag/v0.1.88) will include our "swallowing" patch too.

File: logging/impl.go [hunk 9]
```diff
@@ -32,6 +32,11 @@ type (
 		zapcore.Entry
 		fields []zapcore.Field
 	}
+
+	implWith struct {
```
Comment (benjirewis) [inline logging/impl.go:36]: [q] Why not add a `logFields` field to `impl` and just create a new `impl` object when `WithFields` is called? Or, are we relying on chains of derived `implWith`s to point back to the same `*impl` object and therefore have the same logging level that can be configured by one `log` pattern?

File: logging/impl.go (hunk 6 above)
Comment (benjirewis) [reply logging/impl.go]: > this makes it so that the file name / line number don't properly get logged

This is _probably_ somehow avoidable, but I'm personally fine with the copying. The risk we do run is that we fix some issue in `impl` and forget to fix it in `implWith`.

File: logging/impl.go (hunk 9 above)
Comment (danielbotros) [reply logging/impl.go:36]: > Or, are we relying on chains of derived implWiths to point back to the same *impl object and therefore have the same logging level that can be configured by one log pattern?

yup exactly this! if we return a new `*impl` object we currently have no way of keeping track of it without overwriting the registry entry for the original logger

**Review by benjirewis — APPROVED** [review]
